benchmark/kmeans.py

```diff
@@ -0,0 +1,215 @@
+            )
+            assert_allclose(
+                self.results["cluster_centers"],
+                estimator.cluster_centers_,
+                err_msg=runtime_error_message,
+            )
+            assert_allclose(
+                self.results["inertia"],
+                estimator.inertia_,
+                err_msg=runtime_error_message,
+            )
+
+    def _check_kmeans_single_lloyd_fn_signature(self, kmeans_single_lloyd_fn):
+        fn_signature = inspect.signature(kmeans_single_lloyd_fn)
+        if fn_signature != self._VANILLA_SKLEARN_LLOYD_SIGNATURE:
+            raise ValueError(
+                f"The signature of the submitted kmeans_single_lloyd_fn is expected to be "
+                f"{self._VANILLA_SKLEARN_LLOYD_SIGNATURE}, but got {fn_signature} ."
+            )
+
+
+if __name__ == "__main__":
+    from ext_helpers.sklearn import kmeans as sklearn_kmeans
+    from ext_helpers.daal4py import kmeans as daal4py_kmeans
+
+    from sklearn_numba_dpex.kmeans.drivers import LLoydKMeansDriver
+    from sklearn.datasets import fetch_openml
+    from sklearn.preprocessing import MinMaxScaler
+    from sklearnex import config_context
+    from numpy.random import default_rng
+
+    random_state = 123
+    n_clusters = 127
+    max_iter = 100
+    tol = 0
+    dtype = np.float32
+    # NB: it seems that currently the estimators in the benchmark always return
+    # close results but with significant differences for a few elements.
+    run_consistency_checks = False
+
+    def benchmark_data_initialization(random_state=random_state, n_clusters=n_clusters):
+        X, _ = fetch_openml(name="spoken-arabic-digit", return_X_y=True)
+        X = X.astype(dtype)
+        scaler_x = MinMaxScaler()
+        scaler_x.fit(X)
+        X = scaler_x.transform(X)
+        # Change dataset dimensions
+        # X = np.hstack([X for _ in range(5)])
+        # X = np.vstack([X for _ in range(20)])
+        rng = default_rng(random_state)
+        init = np.array(rng.choice(X, n_clusters, replace=False), dtype=np.float32)
+        return X, None, init
+
+    kmeans_timer = KMeansLloydTimeit(
+        benchmark_data_initialization, max_iter, run_consistency_checks
+    )
+
+    kmeans_timer.timeit(
+        sklearn_kmeans,
+        name="Sklearn vanilla lloyd",
+    )
+
+    with config_context(target_offload="cpu"):
+        kmeans_timer.timeit(
+            daal4py_kmeans,
+            name="daal4py lloyd CPU",
+        )
+
+    with config_context(target_offload="gpu"):
+        kmeans_timer.timeit(
+            daal4py_kmeans,
+            name="daal4py lloyd GPU",
+        )
+
+    for multiplier in [1, 2, 4, 8]:
+        kmeans_timer.timeit(
+            LLoydKMeansDriver(
+                device="cpu", dtype=dtype, work_group_size_multiplier=multiplier
+            ),
+            name=f"Kmeans numba_dpex lloyd CPU (work_group_size_multiplier={multiplier})",
+        )
```

Comment:
Since those are much slower to run, it would be handy to have a command line flag to skip the CPU / numba-dpex combo.